dali/imgcodec/image_decoder.cc

```diff
@@ -0,0 +1,607 @@
+
+    auto &buf_ptr = owner.temp_buffers_[idx];
+    buf_ptr = mm::alloc_raw_async_shared<char, mm::memory_kind::pinned>(
+      size, mm::host_sync, ctx.stream, kTempBufferAlignment);
+
+    SampleView<CPUBackend> &cpu_out = cpu_outputs[i];
+    cpu_out = SampleView<CPUBackend>(buf_ptr.get(), gpu_out.shape(), gpu_out.type());
+  }
+}
+
+void ImageDecoder::DecoderWorker::process_batch(std::unique_ptr<ScheduledWork> work) {
+  auto mask = decoder_->CanDecode(work->ctx,
+                                  make_span(work->sources),
+                                  work->params,
+                                  make_span(work->rois));
+
+  std::unique_ptr<ScheduledWork> fallback_work;
+  if (fallback_) {
+    fallback_work = owner_->new_work(work->ctx, work->results);
+    move_to_fallback(fallback_work.get(), *work, mask);
+    if (!fallback_work->empty())
+      fallback_->add_work(std::move(fallback_work));
+  } else {
+    filter(*work, mask);
+    for (size_t i = 0; i < mask.size(); i++) {
+      if (!mask[i])
+        work->results.set(i, DecodeResult::Failure(nullptr));
+    }
+  }
+
+  if (fallback_work)
+    fallback_->add_work(std::move(fallback_work));
+
+  if (!work->sources.empty()) {
+    bool decode_to_gpu = produces_gpu_output_;
+
+    if (!decode_to_gpu && work->cpu_outputs.empty()) {
+      work->alloc_temp_cpu_outputs(*owner_);
+    }
+
+    auto future = decode_to_gpu
+      ? decoder_->ScheduleDecode(work->ctx,
+                                 make_span(work->gpu_outputs),
+                                 make_span(work->sources),
+                                 work->params,
+                                 make_span(work->rois))
+      : decoder_->ScheduleDecode(work->ctx,
+                                 make_span(work->cpu_outputs),
+                                 make_span(work->sources),
+                                 work->params,
+                                 make_span(work->rois));
+
+    if (fallback_) {
+      for (;;) {
+        auto indices = future.wait_new();
+        if (indices.empty())
+          break;
+
+        auto r = future.get_all_ref();
+
+        for (int sub_idx : indices) {
+          if (r[sub_idx].success) {
+            work->results.set(work->indices[sub_idx], r[sub_idx]);
+          } else {
+            if (!fallback_work)
+              fallback_work = owner_->new_work(work->ctx, work->results);
+            fallback_work->move_entry(*work, sub_idx);
+          }
+        }
+
+        if (fallback_work)
+          fallback_->add_work(std::move(fallback_work));
+      }
+    } else {
+      // no fallback - the results are final
+      future.wait_all();
+      auto r = future.get_all();
+      // we may need to copy to the gpu
+      if (!decode_to_gpu && !work->gpu_outputs.empty()) {
+        for (int i = 0; i < r.size(); i++) {
+          if (r[i].success) {  // TODO(michalz): bulk copy
+            copy(work->gpu_outputs[i], work->cpu_outputs[i], work->ctx.stream);
```

Comment:
This copy will happen only if a fallback was used. As a result, when we successfully decode whole back in the first attempt using CPU-only decoder (libtiff, libjpeg-turbo), no copy to GPU will happen.

We should probably have a copy in the other branch too.

Reply:
Fixed.